File: Week 06/id_276/Leetcode_208.java
class TrieNode {
    char val;
    boolean isWord;
    TrieNode[] chirld = new TrieNode[26];
    TrieNode(){}
    TrieNode(char c) {
        TrieNode node = new TrieNode();
        node.val = c;
    }
}
public class Trie {
    public TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    public void insert(String word) {
        TrieNode next = root;
        for (char c : word.toCharArray()) {
            if (next.chirld[c-'a'] == null) {
                next.chirld[c-'a'] = new TrieNode(c);
            }
            next = next.chirld[c-'a'];
        }
        next.isWord = true;
    }
    public boolean search(String word) {
        TrieNode next = root;
        for (char c : word.toCharArray()) {
            if (next.chirld[c-'a'] == null) return false;
            else next = next.chirld[c-'a'];
        }
        return next.isWord;
    }
    public boolean startsWith(String prefix) {
        TrieNode next = root;
        for (char c : prefix.toCharArray()) {
            if (next.chirld[c-'a'] == null) return false;
            else next = next.chirld[c-'a'];
        }
        return true;
    }
}

File: Week 06/id_381/leetcode-208.py
class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}
        self.end_of_word = '#'
        

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self.root
        for w in word:
            node = node.setdefault(w, {})
        node[self.end_of_word] = self.end_of_word
        

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self.root
        for w in word:
            if w not in node:
                return False
            node = node[w]
        return self.end_of_word in node
        

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.root
        for w in prefix:
            if w not in node:
                return False
            node = node[w]
        return True
